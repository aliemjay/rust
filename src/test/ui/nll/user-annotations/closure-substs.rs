// Test that we enforce user-provided type annotations on closures.

fn foo<'a>() {
    // Here `x` is free in the closure sig:
    |x: &'a i32| -> &'static i32 {
        return x; //~ ERROR lifetime may not live long enough
    };
}

fn foo1() {
    // Here `x` is bound in the closure sig:
    |x: &i32| -> &'static i32 {
        return x; //~ ERROR lifetime may not live long enough
    };
}

fn bar<'a>() {
    // Here `x` is free in the closure sig:
    |x: &'a i32, b: fn(&'static i32)| {
        //~^ ERROR lifetime may not live long enough
        b(x);
    };
}

fn bar1() {
    // Here `x` is bound in the closure sig:
    |x: &i32, b: fn(&'static i32)| {
        //~^ ERROR lifetime may not live long enough
        b(x);
    };
}

fn main() {}
