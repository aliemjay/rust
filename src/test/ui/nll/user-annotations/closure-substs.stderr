error: lifetime may not live long enough
  --> $DIR/closure-substs.rs:6:16
   |
LL | fn foo<'a>() {
   |        -- lifetime `'a` defined here
...
LL |         return x;
   |                ^ returning this value requires that `'a` must outlive `'static`

error: lifetime may not live long enough
  --> $DIR/closure-substs.rs:13:16
   |
LL |     |x: &i32| -> &'static i32 {
   |         - let's call the lifetime of this reference `'1`
LL |         return x;
   |                ^ returning this value requires that `'1` must outlive `'static`

error: lifetime may not live long enough
  --> $DIR/closure-substs.rs:19:18
   |
LL | fn bar<'a>() {
   |        -- lifetime `'a` defined here
LL |     // Here `x` is free in the closure sig:
LL |     |x: &'a i32, b: fn(&'static i32)| {
   |                  ^ type annotation requires that `'a` must outlive `'static`

error: lifetime may not live long enough
  --> $DIR/closure-substs.rs:27:15
   |
LL |     |x: &i32, b: fn(&'static i32)| {
   |         -     ^ type annotation requires that `'1` must outlive `'static`
   |         |
   |         let's call the lifetime of this reference `'1`
LL |
LL |         b(x);
   |         ---- because of argument here

error: aborting due to 4 previous errors

